python/lsst/daf/butler/registry/obscore/_records.py

```diff
@@ -0,0 +1,247 @@
+
+        # add extra columns
+        extra_columns = {}
+        if self.config.extra_columns:
+            extra_columns.update(self.config.extra_columns)
+        if dataset_config.extra_columns:
+            extra_columns.update(dataset_config.extra_columns)
+        for key, column_value in extra_columns.items():
+            # Try to expand the template with known keys, if expansion
+            # fails due to a missing key name then store None.
+            if isinstance(column_value, Mapping):
+                try:
+                    value = column_value["template"].format(**fmt_kws)
+                    record[key] = _TYPE_CONVERSION[column_value["type"]](value)
+                except KeyError:
+                    pass
+            else:
+                # Just a static value.
+                record[key] = column_value
+
+        return record
+
+    def _region_to_columns(self, region: Optional[Region], record: Dict[str, Any]) -> None:
+        if region is None:
+            return
+
+        # Get spatial parameters from the bounding circle.
+        circle = region.getBoundingCircle()
+        center = LonLat(circle.getCenter())
+        record["s_ra"] = center.getLon().asDegrees()
+        record["s_dec"] = center.getLat().asDegrees()
+        record["s_fov"] = circle.getOpeningAngle().asDegrees() * 2
+
+        if isinstance(region, ConvexPolygon):
+            poly = ["POLYGON ICRS"]
+            for vertex in region.getVertices():
+                lon_lat = LonLat(vertex)
+                poly += [
+                    f"{lon_lat.getLon().asDegrees():.6f}",
+                    f"{lon_lat.getLat().asDegrees():.6f}",
+                ]
+            record["s_region"] = " ".join(poly)
+        else:
+            _LOG.warning(f"Unexpected region type: {type(region)}")
+
+    def _exposure_records(self, dimension_record: DimensionRecord, record: Dict[str, Any]) -> None:
+        """Extract all needed info from a visit dimension record."""
+        record["t_exptime"] = dimension_record.exposure_time
+        record["target_name"] = dimension_record.target_name
+
+    def _visit_records(self, dimension_record: DimensionRecord, record: Dict[str, Any]) -> None:
+        """Extract all needed info from an exposure dimension record."""
+        record["t_exptime"] = dimension_record.exposure_time
+        record["target_name"] = dimension_record.target_name
+
+    def _check_collections(self, collection_name: str) -> bool:
+        """Check that specified collection (usually RUN or TAGGED) or any
+        parent CHAINED collection is in the configured collection list.
+        """
+
+        def _all_collections(key: Any) -> Iterator[CollectionRecord]:
+            """Return records of all chained collections."""
+            for collection_record in self.collections.getParentChains(key):
+                yield collection_record
+                yield from _all_collections(collection_record.key)
```

Comment:
I'd certainly prefer not to have logic like this run per-row, as it seems to be doing.  I think whenever we insert new datasets they're grouped by RUN in advance, and of course we also get that kind of grouping when associating into TAGGED collections.  It'd be nice to restructure to take advantage of that.

Reply:
I have updated that logic (it has actually moved to _manager.py in later commit) to reduce the number of checks.